### ISSUE TITLE
<$t:ty> falls victim to C++ >> ambiguity

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#37175]
Consider this macro:

```rust
macro_rules! m { (<$t:ty>) => { stringify!($t) } }
```

Unfortunately it can't deal with a type that ends in `>`:

```rust
m!(<Vec<i32>>)
```

```
error: no rules expected the token `>>`
 --> .cargo/.cargo/script-cache/expr-edcce559d98013ba/expr.rs:5:64
  |
5 | {macro_rules! m { (<$t:ty>) => { stringify!($t) } } m!(<Vec<i32>>)}
  |                                                                ^^

error: Could not compile `expr`.
```

Is it possible to fix this? It seems to me the macro parser should know it's parsing `$t:ty` and therefore `>>` can't be valid, so it should be split into two tokens.

cc @jseyfried

**Comment by jseyfried** [comment]
@durka yeah, this is definitely possible to fix.

The parser already splits the `>>` into two tokens and parses the type correctly -- the problem appears to be that we drop that parser once we finish parsing the type and just keep the underlying `TtReader` (stream of tokens), which doesn't know that the `>>` has been partially consumed (c.f. the call to `parse_nt` in `syntax/ext/tt/macro_parser.rs`).

To fix this, we could to have the parser mutate the `>>` into a `>` in the `TtReader` (currently, we can only consume tokens from a `TtReader`). Or, we could also check if there is a partially consumed token before dropping the parser and keep track of it locally in `macro_parser::parse`.

**Comment by jseyfried** [comment]
Fixed in #37208.
